crypto/src/hasher/poseidon.rs

```diff
@@ -0,0 +1,98 @@
+//! Mina Poseidon hasher
+//!
+//! An implementation of Mina's hasher based on the poseidon arithmetic sponge
+//!
+use std::marker::PhantomData;
+
+use crate::hasher::DomainParameter;
+use mina_curves::pasta::Fp;
+use oracle::{
+    pasta,
+    poseidon::{
+        ArithmeticSponge, ArithmeticSpongeParams, PlonkSpongeConstantsKimchi,
+        PlonkSpongeConstantsLegacy, Sponge, SpongeConstants, SpongeState,
+    },
+};
+
+use super::{domain_prefix_to_field, Hashable, Hasher};
+
+/// There poseidon hasher structure
+//
+//  The arithmetic sponge parameters are large and costly to initialize,
+//  so we only want to do this once and then re-use the Poseidon context
+//  for many hashes. Also, following approach of the mina code we store
+//  a backup of the initialized sponge state for efficient reuse.
+pub struct Poseidon<SC: SpongeConstants, H: Hashable> {
+    sponge: ArithmeticSponge<Fp, SC>,
+    sponge_state: SpongeState,
+    state: Vec<Fp>,
+    phantom: PhantomData<H>,
+}
+
+impl<SC: SpongeConstants, H: Hashable> Poseidon<SC, H> {
+    fn new(domain_param: H::D, sponge_params: ArithmeticSpongeParams<Fp>) -> Self {
+        let mut poseidon = Poseidon::<SC, H> {
+            sponge: ArithmeticSponge::<Fp, SC>::new(sponge_params),
+            sponge_state: SpongeState::Absorbed(0),
+            state: vec![],
+            phantom: PhantomData,
+        };
+
+        poseidon.init(domain_param);
+
+        poseidon
+    }
+}
+
+pub(crate) fn new_legacy<H: Hashable>(domain_param: H::D) -> impl Hasher<H> {
```

Comment:
We don't impose a particular return type here as opposed to `new_kimchi` ?

Reply:
Thanks-- I'll change to be consistent.  👍🏼